### DISCUSSION TITLE
Software PWM minimal frequency

### DISCUSSION THREAD

**Discussion by msawicki-poland** [Pi4J/pi4j#293] [Q&A]
I use 2.3.0 pi4j library and Kotlin and I created pwm config like this:
```
fun pwmConfig(pi4j: Context): PwmConfig = Pwm.newConfigBuilder(pi4j)
        .address(27)
        .pwmType(PwmType.SOFTWARE)
        .provider("pigpio-pwm")
        .initial(0)
        .shutdown(0)
        .build()
```
and I run pwm like this:
`pwm.on(4.0, 2)`
On osciloscope I found that it generates 10Hz not 2Hz. Why I cannot run pwm for 1-9Hz?

**Comment by taartspi** [comment]
PWM software uses the Pigpio library    https://abyz.me.uk/rpi/pigpio/index.html        That on() method calls gpioSetPWMfrequency in that library.  Their doc has a table showing possible frequencies.   That details says the 'sample rate' is used to determine the chart value defining minimum freq, apparently the sample rate resulted as 10 the lowest possible .  I tested freq of 18 results in the actual freq being 20.   I do not see that we support setting the sample rate (gpioCfgClock)                 All that said, we already recognize the desired software frequency may not be possible :                         
    /**
     * Turn the PWM signal [ON] using a specified duty-cycle (%)
     * at the pre-configured frequency (Hz).
     *
     * @param dutyCycle  The duty-cycle value is a decimal value that represents the
     *                   percentage of the ON vs OFF time of the PWM signal for each
     *                   period.  A value of 50 represents a duty-cycle where half of
     *                   the time period the signal is LOW and the other half is HIGH.
     *                   The duty-cycle range is valid from 0 to 100 including factional
     *                   values.  (Values above 50% mean the signal will remain HIGH more
     *                   time than LOW.)
     * @param frequency  The desired frequency value in Hertz (number of cycles per second)
     *                   that the PWM signal generator should attempt to output.  Please
     *                   note that certain PWM signal generators may be limited to
     *                   specific frequency bands and may not generate all possible explicit
     *                   frequency values.  Immediately after calling this method, you can
     *                   check the 'Pwm::actualFrequency()' or 'Pwm::getActualFrequency()'
     *                   properties to determine what frequency the PWM generator actually
     *                   applied.
     * @return returns this PWM instance
     * @throws IOException if fails to communicate with the PWM pin
     */                                                                                                                                                                                                                           
                     This info is in the java docs, I will ask this question be converted to a work item and update the web pages with this caution.    @FDelporte    Sir, can you convert this to a DOC work item and I will update the web pages to include this caution.

**Comment by BetterB0y** [comment]
So settings even 8kHz isn't possible, because of lack of sample rate setter?

**Comment by taartspi** [comment]
i assume you mean 8 hz, not 8KHz       Correct, appears 8 HZ will have an actual freq of 10 Hz,    If you can spare using a hardware PWM it appears to do what you want.  I didn't connect a my scanaquad to watch the line but looks like the LED is flashing at that slow rate.

**Comment by BetterB0y** [comment]
I meant 8 kHZ. We only achived max 2 kHZ. I think in documentation we found that max we can do 8 kHZ of software pwm

**Comment by taartspi** [comment]
your test case used 2 Hz, I did not realize you were starting a different question.    I would suggest use the step  written in the javaDoc to set the freq and inspect its actual value by calling the method'Pwm::getActualFrequency()     If actual is 8Khz and signal measurement is far less, I suspect this is what the web pages mentions that soft pwm is not accurate compared to hw pwm.           "software version cannot achieve precise or particularly fast frequencies."
